sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -502,11 +502,11 @@ pi_result piProgramCreate(pi_context context, const void *il, size_t length,
   }
 
   size_t extSize;
-  ret_err = clGetPlatformInfo(curPlatform, CL_PLATFORM_EXTENSIONS, 0, nullptr,
-                              &extSize);
+  ret_err = clGetDeviceInfo(cast<cl_device_id>(device), CL_DEVICE_EXTENSIONS,
```

Comment:
I think it would be more clear if the check is done in the SYCL level rather than in the plugins. Doing so also would not require changing PI API.

Tagging @smaslov-intel

Reply:
I agree with Vlad that we'd better check device's "cl_khr_il_program" extension outside of `piProgramCreate`.
One possible place is to do that in `piProgramBuild` where the list of devices is known,

Reply:
Thanks for for having a look!

@smaslov-intel Depending on the device version and whether `cl_khr_il_program` is supported by the device we either need to call `clCreateProgramWithIL()` or `clCreateProgramWithILKHR()`, hence we need to check device version and extensions either within the OpenCL plugin's `piProgramCreate()` or somewhen before and store the particular function pointer somewhere for later usage. So I guess `piProgramBuild()` wouldn't work.

@romanovvlad What do you mean with "doing it in the SYCL level rather than than in the plugins"? Since those checks are OpenCL specific, isn't the OpenCL plugin the only place it belongs to? Also, as mentioned above it directly influences whether to call `clCreateProgramWithIL()` or `clCreateProgramWithILKHR()` in the OpenCL plugin's `piProgramCreate()`.

Reply:
Is it possible that not all devices in the OpenCL context support the same way of creating a program on that context? Why not just check first (all) devices then?

Reply:
I think they can all be in the same context. The specification seems to keep it open, but in order to make it work implementations must compute the smallest common denominator of what all devices in the context support and stick to that when calling `clCreateProgramWithIL*()` and fail if not all devices are able to deal with the IL type and version. If one device does not support it at all the call should just fail as well.

Anyway, I misread your previous comment, just checking all devices in the context should work and would even be necessary if there'd be multiple ones associated with the context. I indeed missed that - yikes!

I'll get rid of the PI API change, fix up the other patch and update the branch.